src/hugr/rewrite.rs

```diff
@@ -0,0 +1,57 @@
+//! Replace operations on the HUGR.
+
+#[allow(clippy::module_inception)] // TODO: Rename?
+pub mod replace;
+use std::mem;
+
+use crate::Hugr;
+pub use replace::{OpenHugr, Replace, ReplaceError};
+
+/// An operation that can be applied to mutate a Hugr
+pub trait Rewrite<E> {
+    /// If `true`, [self.apply]'s of this rewrite guarantee that they do not mutate the Hugr when they return an Err.
+    /// If `false`, there is no guarantee; the Hugr should be assumed invalid when Err is returned.
+    const UNCHANGED_ON_FAILURE: bool;
+
+    /// Checks whether the rewrite would succeed on the specified Hugr.
+    /// If this call succeeds, [self.apply] should also succeed on the same `h`
+    /// If this calls fails, [self.apply] would fail with the same error.
+    fn verify(&self, h: &Hugr) -> Result<(), E>;
+
+    /// Mutate the specified Hugr, or fail with an error.
+    /// If [self.unchanged_on_failure] is true, then `h` must be unchanged if Err is returned.
+    /// See also [self.verify]
+    /// # Panics
+    /// May panic if-and-only-if `h` would have failed [Hugr::validate]; that is,
+    /// implementations may begin with `assert!(h.validate())`, and *should* begin with `debug_assert!(h.validate())`
```

Comment:
@ss2165 this feels a bit strong. `are encouraged to use debug_assert instead` (i.e. if they want to do anything)?

Reply:
yeah I think that works, or somewhat equivalently "may begin with `debug_assert!(h.validate())`"